File: web/src/types/dispatch.ts
export type OfficerStatus = 'AVAILABLE' | 'ASSIGNED' | 'BUSY' | 'OFFLINE';
export type LocationStatus = 'LOCATION_CURRENT' | 'LOCATION_STALE' | 'LOCATION_UNKNOWN';

export type DispatchStatus =
  | 'UNASSIGNED'
  | 'ASSIGNED'
  | 'ACKNOWLEDGED'
  | 'EN_ROUTE'
  | 'ON_SCENE'
  | 'RESPONDING'
  | 'COMPLETED'
  | 'CANCELLED';

export interface ResponseOfficer {
  officer_id: string;
  user_id?: string;
  name: string;
  role: string;
  status: OfficerStatus;
  current_latitude?: number;
  current_longitude?: number;
  location_status: LocationStatus;
  location_timestamp?: string;
  assigned_event_id: string;
  created_at: string;
  updated_at: string;
}

export interface DispatchTransition {
  transition_id: string;
  dispatch_id: string;
  previous_status: DispatchStatus;
  new_status: DispatchStatus;
  timestamp: string;
  actor_type: 'SYSTEM' | 'OPERATOR' | 'FIELD_OFFICER';
  actor_id?: string;
  reason?: string;
  metadata_json?: Record<string, any>;
}

export interface DispatchAssignment {
  dispatch_id: string;
  incident_id: string;
  event_id: string;
  officer_id: string;
  status: DispatchStatus;
  assigned_by: string;
  assigned_at: string;
  acknowledged_at?: string;
  en_route_at?: string;
  on_scene_at?: string;
  responding_at?: string;
  completed_at?: string;
  cancelled_at?: string;
  eta_minutes?: number;
  dispatch_reason?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
  officer?: ResponseOfficer;
  transitions?: DispatchTransition[];
}

export interface FieldOfficerAssignmentContext {
  dispatch: DispatchAssignment;
  incident_id: string;
  zone_id: string;
  event_id: string;
  camera_id?: string;
  warning_state: string;
  physics_risk?: number;
  ai_probability?: number;
  model_version: string;
  label_type: string;
  model_status: string;
  disclaimer: string;
}

export const VALID_DISPATCH_TRANSITIONS: Record<DispatchStatus, DispatchStatus[]> = {
  UNASSIGNED: ['ASSIGNED'],
  ASSIGNED: ['ACKNOWLEDGED', 'CANCELLED'],
  ACKNOWLEDGED: ['EN_ROUTE', 'CANCELLED'],
  EN_ROUTE: ['ON_SCENE', 'CANCELLED'],
  ON_SCENE: ['RESPONDING', 'CANCELLED'],
  RESPONDING: ['COMPLETED', 'CANCELLED'],
  COMPLETED: [],
  CANCELLED: [],
};

File: backend/app/schemas/dispatch.py
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class ResponseOfficerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    officer_id: str
    user_id: Optional[str] = None
    name: str
    role: str = "FIELD_OFFICER"
    status: str = "AVAILABLE"  # AVAILABLE, ASSIGNED, BUSY, OFFLINE
    current_latitude: Optional[float] = None
    current_longitude: Optional[float] = None
    location_status: str = "LOCATION_UNKNOWN"  # LOCATION_CURRENT, LOCATION_STALE, LOCATION_UNKNOWN
    location_timestamp: Optional[datetime] = None
    assigned_event_id: str = "evt_01"
    created_at: datetime
    updated_at: datetime


class DispatchCreateRequest(BaseModel):
    officer_id: str = Field(..., description="ID of the field officer to assign")
    eta_minutes: Optional[int] = Field(default=5, ge=1, le=180, description="Estimated time of arrival in minutes")
    reason: str = Field(..., min_length=3, description="Mandatory dispatch operational justification reason")


class DispatchTransitionRequest(BaseModel):
    new_status: str = Field(..., description="Target status: ACKNOWLEDGED, EN_ROUTE, ON_SCENE, RESPONDING, COMPLETED, CANCELLED")
    reason: Optional[str] = Field(None, description="Optional operational notes or reason for status transition")


class DispatchTransitionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    transition_id: str
    dispatch_id: str
    previous_status: str
    new_status: str
    timestamp: datetime
    actor_type: str
    actor_id: Optional[str] = None
    reason: Optional[str] = None
    metadata_json: Optional[Dict[str, Any]] = None


class DispatchCanonicalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    dispatch_id: str
    incident_id: str
    event_id: str
    officer_id: str
    status: str
    assigned_by: str
    assigned_at: datetime
    acknowledged_at: Optional[datetime] = None
    en_route_at: Optional[datetime] = None
    on_scene_at: Optional[datetime] = None
    responding_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    cancelled_at: Optional[datetime] = None
    eta_minutes: Optional[int] = None
    dispatch_reason: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    officer: Optional[ResponseOfficerResponse] = None
    transitions: List[DispatchTransitionResponse] = []


class FieldOfficerAssignmentContextResponse(BaseModel):
    dispatch: DispatchCanonicalResponse
    incident_id: str
    zone_id: str
    event_id: str
    camera_id: Optional[str] = None
    warning_state: str
    physics_risk: Optional[float] = None
    ai_probability: Optional[float] = None
    model_version: str = "v2.0.0"
    label_type: str = "PHYSICS_DEFINED_PROXY"
    model_status: str = "PROTOTYPE"
    disclaimer: str = "AI Early Warning — Prototype. Physics-defined proxy model, not operationally validated."
